[core][state] Task events backend - gcs update path and interface [2/n]

rickyyx opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?


Previous PRs:
 - https://github.com/ray-project/ray/pull/30829: Interface and protobuf definitions. 
 - https://github.com/ray-project/ray/pull/30953: Splitting of drop count for various events type on worker.  
 - https://github.com/ray-project/ray/pull/30867: `TaskEventBuffer` implementation 


In this PR: 
- Added `GcsTaskManager` that stores the task events on the GCS side. 
- The `GcsTsakManager` has its own io service and io thread that's separated from the main rpc thread/io_context. Handling of rpcs will be posted to its own internal io_service. 
- Implementation for the update path. 
- Interface for the read path. 


Next PRs: 
- Implementation for the update path of `GcsTaskManager`
- Porting of profiling events 
- Porting of state api task. 

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.h:
```diff
@@ -0,0 +1,161 @@
+    RAY_LOG(FATAL) << "Unimplemented!";
+  };
+
+  /// Stops the event loop and the thread of the task event handler.
```
```suggestion
  /// Stops the event loop and the thread of the task event handler.
  /// After this API is called, it cannot handle any requests 
```

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.h:
```diff
@@ -0,0 +1,161 @@
+  /// Returns the io_service.
+  ///
+  /// \return Reference to its io_service.
+  instrumented_io_context &GetIoContext() { return io_service_; }
```
const instrumented_io_context &GetIoContext() const { return io_service_; }

rickyyx replied:
hmm - this needs to be an non-const reference since the [caller](https://github.com/ray-project/ray/blob/561344a084d3e30622b1e2fa560d69755a5ebbf1/src/ray/rpc/grpc_server.h#L151) takes a non-const reference.

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+                                            rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(DEBUG) << "Adding task state event:" << request.data().ShortDebugString();
+  // Dispatch to the handler
+  io_service_.post(
```
Is this necessary? I think this method should be dispatched to this io service automatically?

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+          auto task_id = TaskID::FromBinary(events_by_task.task_id());
+          // TODO(rickyx): add logic to handle too many profile events for a single task
+          // attempt.
+          auto replaced_task_events =
```
does this avoid copy?

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+          }
+          RAY_LOG(DEBUG) << "Processed a task event. [task_id=" << task_id.Hex() << "]";
+        }
+        RAY_LOG_EVERY_MS(INFO, 10000)
```
maybe just adding this into debug_state_gcs.txt instead of periodically printing it?

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+absl::optional<rpc::TaskEvents>
+GcsTaskManager::GcsTaskManagerStorage::AddOrReplaceTaskEvent(
+    rpc::TaskEvents events_by_task) {
+  RAY_LOG_EVERY_MS(INFO, 10000)
```
same comments as below (use debug state?)

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+
+    // Change the underlying storage.
+    auto &to_replaced = task_events_.at(next_idx_to_overwrite_);
+    std::swap(to_replaced, events_by_task);
+    auto replaced = events_by_task;
```
```suggestion
    std::swap(to_replaced, events_by_task);
    auto replaced = std::move(events_by_task);
```
?

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+  }
+}
+
+absl::optional<rpc::TaskEvents>
```
Consider returning bool? I think the dropped counter can be tracked within storage?

rickyyx replied:
I feel the drop count could actually be tracked outside of the storage, such that the storage could be better decoupled. But open to move the tracking in.

rkooo567 commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+
+absl::optional<rpc::TaskEvents>
+GcsTaskManager::GcsTaskManagerStorage::AddOrReplaceTaskEvent(
+    rpc::TaskEvents events_by_task) {
```
```suggestion
    rpc::TaskEvents &events_by_task) {
```

rkooo567 reviewed (commented):
Let me have additional look later.

rickyyx commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+
+          if (replaced_task_events) {
+            if (replaced_task_events->has_state_updates()) {
+              // TODO(rickyx): should we un-flatten the status updates into a list of
```
This is actually independent from if we do client side batching. The replaced message here is already merged at gcs. With profile events, we are able to know the number of actual profile events being evicted, but for status changes, we don't. 

Or I guess we could check the number of fields being set in the state updates.

rickyyx commented on src/ray/gcs/gcs_server/gcs_task_manager.cc:
```diff
@@ -0,0 +1,146 @@
+    auto &existing_events = task_events_.at(idx);
+
+    // Update the events.
+    num_bytes_task_events_ -= existing_events.ByteSizeLong();
```
Right - I agree, but so far this has not been showing up in the perf graphs so I thought i would just keep them. I did have experimenets before where I removed it and didn't see performance gain in microbenchmark.

ericl commented on src/ray/gcs/gcs_server/gcs_task_manager.h:
```diff
@@ -0,0 +1,161 @@
+    const size_t max_num_task_events_ = 0;
+
+    /// Current task events stored.
+    std::vector<rpc::TaskEvents> task_events_;
```
Fine for now, but if the logic gets more complicated we should consider splitting this out into a LRU data struture.

ericl reviewed (commented):
Nice test! Consider adding (1) DebugString() to expose in debug state, and also (2) prometheus gauges for the count/bytes added / currently stored in the task stats storage.

rickyyx commented:
Addressed most of the comments: 
- Moved debug states printing to `DebugString`
- Added metric reporting for various counters 
- Added e2e integration test for metric export
- Removed unnecessary post on io_context 
- Removed Gcs read rpc definitions for next PRs. 

PTAL @rkooo567 @ericl

ericl commented on python/ray/tests/test_task_events.py:
```diff
@@ -0,0 +1,53 @@
+def count_by_type(info: RayContext) -> Dict:
+    res = raw_metrics(info)
+    task_events_info = defaultdict(int)
+    if "ray_gcs_task_manager_task_events_count" in res:
```
nit: typically you'd use two separate metrics in this case, since COUNT vs STORED are not summable (don't add to 100%).

ericl reviewed (approved):
Looks good, but consider splitting into separate metrics for reported, stored, dropped.
